## Условие 1: 1 задача
Скачать файл в уроке  
Считать данные с помощью pandas  
Вывести на экран первые 5 строк  
Посмотреть на описание признаков и на их содержание

In [53]:
import pandas as pd


df = pd.read_csv("kc_final.csv")
df.head(5)

Unnamed: 0          id        date     price  bedrooms  bathrooms  \
0           0  7129300520  2014-10-13  221900.0         3       1.00   
1           1  6414100192  2014-12-09  538000.0         3       2.25   
2           2  5631500400  2015-02-25  180000.0         2       1.00   
3           3  2487200875  2014-12-09  604000.0         4       3.00   
4           4  1954400510  2015-02-18  510000.0         3       2.00   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1180      5650     1.0           0  ...      7        1180   
1         2570      7242     2.0           0  ...      7        2170   
2          770     10000     1.0           0  ...      6         770   
3         1960      5000     1.0           0  ...      7        1050   
4         1680      8080     1.0           0  ...      8        1680   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 22 columns]

## Условие 2: 2 задача
Проведите первичный анализ данных  
Изучите типы данных  
Найдите количество пропущенных ячеек в данных  
Посчитайте основные статистики по всем признакам и поизучайте их  
Пишите выводы

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     21613 non-null  int64  
 1   id             21613 non-null  int64  
 2   date           21613 non-null  object 
 3   price          21613 non-null  float64
 4   bedrooms       21613 non-null  int64  
 5   bathrooms      21613 non-null  float64
 6   sqft_living    21613 non-null  int64  
 7   sqft_lot       21613 non-null  int64  
 8   floors         21613 non-null  float64
 9   waterfront     21613 non-null  int64  
 10  view           21613 non-null  int64  
 11  condition      21613 non-null  int64  
 12  grade          21613 non-null  int64  
 13  sqft_above     21613 non-null  int64  
 14  sqft_basement  21613 non-null  int64  
 15  yr_built       21613 non-null  int64  
 16  yr_renovated   21613 non-null  int64  
 17  zipcode        21613 non-null  int64  
 18  lat   

In [4]:
df.isnull().sum()

Unnamed: 0       0
id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [50]:
df.describe()

Unnamed: 0            id         price      bedrooms     bathrooms  \
count  21613.00000  2.161300e+04  2.161300e+04  21613.000000  21613.000000   
mean   10806.00000  4.580302e+09  5.400881e+05      3.370842      2.114757   
std     6239.28002  2.876566e+09  3.671272e+05      0.930062      0.770163   
min        0.00000  1.000102e+06  7.500000e+04      0.000000      0.000000   
25%     5403.00000  2.123049e+09  3.219500e+05      3.000000      1.750000   
50%    10806.00000  3.904930e+09  4.500000e+05      3.000000      2.250000   
75%    16209.00000  7.308900e+09  6.450000e+05      4.000000      2.500000   
max    21612.00000  9.900000e+09  7.700000e+06     33.000000      8.000000   

        sqft_living      sqft_lot        floors    waterfront          view  \
count  21613.000000  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean    2079.899736  1.510697e+04      1.494309      0.007542      0.234303   
std      918.440897  4.142051e+04      0.539989      0.086517      0.766318   
min      290.000000  5.200000e+02      1.000000      0.000000      0.000000   
25%     1427.000000  5.040000e+03      1.000000      0.000000      0.000000   
50%     1910.000000  7.618000e+03      1.500000      0.000000      0.000000   
75%     2550.000000  1.068800e+04      2.000000      0.000000      0.000000   
max    13540.000000  1.651359e+06      3.500000      1.000000      4.000000   

       ...    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  ...  21613.000000   21613.000000  21613.000000  21613.000000   
mean   ...   1788.390691     291.509045   1971.005136     84.402258   
std    ...    828.090978     442.575043     29.373411    401.679240   
min    ...    290.000000       0.000000   1900.000000      0.000000   
25%    ...   1190.000000       0.000000   1951.000000      0.000000   
50%    ...   1560.000000       0.000000   1975.000000      0.000000   
75%    ...   2210.000000     560.000000   1997.000000      0.000000   
max    ...   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  \
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000   
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652   
std       53.505026      0.138564      0.140828     685.391304   27304.179631   
min    98001.000000     47.155900   -122.519000     399.000000     651.000000   
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000   
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000   
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000   
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000   

       living_sqrt_percent  
count         21613.000000  
mean              0.875498  
std               0.170969  
min               0.333333  
25%               0.725979  
50%               1.000000  
75%               1.000000  
max               1.000000  

[8 rows x 22 columns]

#### Выводы:
Таблица представляет собой статистику об объектах недвижимости с характеристиками объекта и стоимостью  
21613 записей  
Значений null нет

## Условие 3: 3 задача
Ответьте на несколько вопросов  

### 3.1 В каком диапазоне изменяются стоимости недвижимости?

In [7]:
(df['price'].min(), df['price'].max())

(75000.0, 7700000.0)

### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

> *Возможно, что это неверный расчет, но другого варианта я не обнаружил, т.к. нет явных переменных а-ля жилая и общая площадь*

In [33]:
df['living_sqrt_percent'] = df['sqft_above']/df['sqft_living']
print(f"Средний процент площади без подвала составялет {df['living_sqrt_percent'].mean().round(2)*100} %")

Средний процент площади без подвала составялет 88.0 %


### 3.3 Как много домов с разными этажами в данных?

> *Тут не совсем понятно что требуется, пришлось выводить через groupby все варианты*

In [44]:
df.groupby(['floors']).size()

floors
1.0    10680
1.5     1910
2.0     8241
2.5      161
3.0      613
3.5        8
dtype: int64

### 3.4 Насколько хорошие состояния у домов в данных?

In [28]:
print(f"Средняя оценка состояния домов = {df['condition'].mean().round(2)} из {df['condition'].max()}")

Средняя оценка состояния домов = 3.41 из 5


### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [30]:
print (f"Самый старый дом {df['yr_built'].min()} года постройки, самый новый - {df['yr_built'].max()}")

Самый старый дом 1900 года постройки, самый новый - 2015


## Условие 4: 4 задача
Ответьте на несколько вопросов

### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [36]:
avg_price = df[df['bedrooms'] == 2]['price'].mean().round(2)
print(f"Средняя цена дома с 2мя спальнями = {avg_price}")

Средняя цена дома с 2мя спальнями = 401372.68


### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [38]:
avg_sqrt = df[df['price'] > 600000]['sqft_living'].mean().round(2)
print(f"Средняя площадь дома с ценой более 600тыс = {avg_sqrt}")

Средняя площадь дома с ценой более 600тыс = 2895.47


### 4.3 Как много домов коснулся ремонт?

In [42]:
cnt_renovated = df[df['yr_renovated'] > 0].shape[0]
print(f"Количество домов после ремонта = {cnt_renovated}")

Количество домов после ремонта = 914


### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [51]:
avg_above_10 = df[df['grade'] > 10]['price'].mean().round(2)
avg_below_4 = df[df['grade'] < 4]['price'].mean().round(2)
diff = avg_above_10 - avg_below_4
print(f"Средняя разница м-ду недвигой с оценкой 10+ и оценкой 4- составляет {diff}")

Средняя разница м-ду недвигой с оценкой 10+ и оценкой 4- составляет 1488885.12


## Условие 5: 5 задача
Ответьте на несколько вопросов
### 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [64]:
variants = df[(df['waterfront'] == 1)&(df['bathrooms'] >= 3)&(df['sqft_basement'] > 0)]
print(f"Количество подходящих вариантов = {variants.shape[0]}")

Количество подходящих вариантов = 41


###5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [63]:
variants = df[((df['view'] == df['view'].max()) | (df['waterfront'] == 1))&(df['condition'] == df['condition'].max())&(df['yr_built'] >= 1980)]
print (f"Подходящие варианты будут в диапазоне от {variants['price'].min()} до {variants['price'].max()}")

Подходящие варианты будут в диапазоне от 1295000.0 до 3000000.0


### 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [85]:
variants = df[(df['sqft_basement'] == 0)&(df['floors'] == 2)&(df['price'] <= 150000)]
print (f"Состояние подходящих вариантов в среднем {variants['condition'].mean()} из {df['condition'].max()}")

Состояние подходящих вариантов в среднем 3.0 из 5
